### PR TITLE
ENH: make check-{docs,tutorials} fail on dtype mismatch

### LINKED ISSUE #27263: TEST: Doctest is not strict enough
Over at #27257 we discovered that the new doctest runner no longer checks the string formatting, it uses `np.allcose()` to verify the results. At the least, it should be checking the dtype is correct, and maybe some other things like shape? cc @ev-br

### PR DESCRIPTION
**Pull request by ev-br**
closes https://github.com/numpy/numpy/issues/27263 

Make the `check-docs` check stricter and fail on dtype mismatch. Previously, it relied on however numpy defines equality, e.g.  `np.float64(3) == 3` hence the doctest below was passing:

```
>>> np.float64(3)
3
```

EDIT: <s>Making the checker stricter needs an a small update in `scipy-doctest`, https://github.com/scipy/scipy_doctest/pull/173, so this PR temporarily installs `scipy-doctest` from a branch, if only to flush the remaining ~100 failures.</s> scipy-doctest==1.4 is up on pypi, so use that. Locally need to do `pip install --upgrade scipy-doctest`

### CODE REVIEW

**Comment by ev-br** [comment]
The remaining CI failure is fixed by https://github.com/numpy/numpy/pull/27262/

**Comment by ngoldbaum** [comment]
close/reopen

**Comment by ngoldbaum** [comment]
Oh wait that won't work, you'd need to merge or rebase.

**Review by ngoldbaum — APPROVED** [review]
Nice! I think these are all clear improvements with the changed scalar reprs. Now users won't be confused anymore why their repr is different from the docs.

**Comment by seberg** [comment]
Changes look good, thanks.  I think this is still "strict" in the sense that it does allow to print a normal integer/float value when it is the default one.

If we make it strict to include the type information there as well, the changes would be much larger (I had the tool to fix this with pytest-doctestplus, but not sure how much work it is to fix unrelated doctestplus things, although one should be able to ignore them for this purpose... Anyway, possible, but a bit of a fiddly task probably).

Anyway, long story short: I think this is great, even if it would be nice to cover the default scalar types as well (although they are platform specific...).

**Comment by ev-br** [comment]
> not sure how much work it is to fix unrelated doctestplus things, ... Anyway, possible, but a bit of a fiddly task probably).

I'm going to go on record with a guess that the amount of fiddling is going to be \approx the same for scipy-doctest or pytest-doctestplus, xdoctest or whatever other tool. The fiddle is not in the checker tech complexity (which may or may not differ), it's just adapting to the needs of the existing body of the docs.

**Comment by seberg** [comment]
The tool I mean is to auto-update the docs based on the actual output (slightly tricky for NumPy because C-docs are hard to find).
But yes, adding that to scipy-doctest is probably also easy, what tool to use doesn't matter much in the end.

**Comment by ev-br** [comment]
Flagging them is not too difficult indeed: https://github.com/scipy/scipy_doctest/issues/60#issuecomment-2308401695
The vast majority is indeed in `add_newdocs.py` variants, and no, I'm not volunteering to fix 537 failures, script or manual :-).
